### PR TITLE
feat(worker): messsage handling

### PR DESCRIPTION
**Pull request by lemosep**
In this PR:
* Implemented `Sender` struct, responsible for making HTTP requests to `ctl::http` component.
* Implemented instance termination with shutdown.
* Refactored worker's main, `Runner` and `ContainerRuntime`

### CODE REVIEW

**Comment by lemosep** [comment]
My approach with `Sender` was based on modularizing a component, so it is implemented both in `monitor` and `runner` modules. This also required me to wrap the struct around a `Arc`, to share among those components properly

File: worker/src/runner/container_rt.rs [hunk 1]
```diff
@@ -31,23 +35,45 @@ impl ContainerRuntime {
...
+            handle
+                .report_instance_status(spec.instance_id, Status::Terminated)
+                .await;
+        });
+    }
+
+    pub fn terminate_instance(&self, id: InstanceId) {
+        let this = self.clone();
+        tokio::spawn(async move {
+            let name = format!("instance-{id:?}");
```
Comment (lffg) [inline worker/src/runner/container_rt.rs]:
```suggestion
            let name = format!("instance-{id}");
```
We shouldn't use debug formatting for this purpose. In general, they aren't stable.

File: worker/src/runner/container_rt.rs [hunk 2]
```diff
@@ -31,23 +35,45 @@ impl ContainerRuntime {
...
                 handle
-                    .report_instance_status(Status::FailedToStart { error })
+                    .report_instance_status(spec.instance_id, Status::FailedToStart { error })
                     .await;
                 return;
             }
 
-            // healthcheck verifies if service is running on established `PORT`
-            handle.report_instance_status(Status::Started).await;
+            // Healthcheck verifies if service is running on established `PORT`
```
Comment (lffg) [inline worker/src/runner/container_rt.rs]:
```suggestion
            // TODO: Add health check to verify whether the service is running
```

File: worker/src/monitor/pusher.rs [hunk 3]
```diff
@@ -1,23 +1,14 @@
...
     loop {
         sleep(args.metrics_report_interval).await;
         let metrics = metrics_report.get_metrics();
-
-        let _ = client
-            .post("http://localhost:3000/worker/metrics")
-            .json(&metrics)
-            .send()
-            .await;
+        let _ = sender.send_metrics(metrics).await;
```
Comment (lffg) [inline worker/src/monitor/pusher.rs]: We shouldn't ignore handling errors, which _will_ happen

File: worker/src/sender/mod.rs [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+#[derive(Clone, Debug)]
+pub struct Sender {
+    pub client: Client,
+    pub http_path: SocketAddr,
+}
+
+impl Sender {
+    /// Creates a `Sender` instance wrapped by an `Arc`.
+    #[must_use]
+    pub fn new(path: SocketAddr) -> Arc<Self> {
```
Comment (lffg) [inline worker/src/sender/mod.rs]: In general, `new` methods return `Self`. This implementation forces downstream users to use an `Arc`, which is not always desirable

File: worker/src/sender/mod.rs [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+        Arc::new(Sender {
+            client: Client::new(),
+            http_path: path,
+        })
+    }
+    /// The `ctl::http` URL should be known by the worker since the start.
+    /// So it should be passed via `args` to the main function.
```
Comment (lffg) [inline worker/src/sender/mod.rs]: This comment is not related to the `send_status` method

File: worker/src/runner/container_rt.rs [hunk 6]
```diff
@@ -31,23 +35,45 @@ impl ContainerRuntime {
...
+            let name = format!("instance-{id:?}");
+            loop {
+                match this.stop_container(&name).await {
+                    Ok(()) => {
+                        info!("Successfully stopped container!");
+                        break;
+                    }
+                    Err(e) => {
+                        info!("Failed to stop container: {:?}. Retrying...", e);
+                    }
```
Comment (lffg) [inline worker/src/runner/container_rt.rs]: What would happen if, e.g., one passes an `InstanceId` which doesn't exist?

File: worker/src/runner/container_rt.rs [hunk 7]
```diff
@@ -115,6 +141,13 @@ impl ContainerRuntime {
         }
     }
 
+    async fn stop_container(&self, name: &str) -> eyre::Result<()> {
+        Ok(self
+            .docker
+            .stop_container(name, None::<StopContainerOptions>)
+            .await?)
```
Comment (lffg) [inline worker/src/runner/container_rt.rs]:
```suggestion
        self.docker
            .stop_container(name, None::<StopContainerOptions>)
            .await?;
        Ok(())
```

File: worker/src/runner/mod.rs [hunk 8]
```diff
@@ -15,17 +15,20 @@ use tokio::{
...
 
+use super::sender;
+
 pub struct Runner {
     rx: mpsc::Receiver<Msg>,
     instances: HashMap<InstanceId, u16>,
     ports: HashSet<u16>,
     handle: RunnerHandle,
     container_runtime: ContainerRuntime,
+    http_sender: Arc<sender::Sender>,
```
Comment (lffg) [inline worker/src/runner/mod.rs]:
```suggestion
    ctl_sender: Arc<sender::Sender>,
```

**Review by lffg — CHANGES_REQUESTED** [review]

File: worker/src/monitor/pusher.rs (hunk 3 above)
Comment (lffg) [reply worker/src/monitor/pusher.rs]: I mean, we should at least log them using `tracing::error` or something like that

File: worker/src/runner/container_rt.rs [hunk 9]
```diff
@@ -115,6 +141,13 @@ impl ContainerRuntime {
         }
     }
 
+    async fn stop_container(&self, name: &str) -> eyre::Result<()> {
+        Ok(self
+            .docker
+            .stop_container(name, None::<StopContainerOptions>)
```
Comment (lffg) [inline worker/src/runner/container_rt.rs]: Termination is a bit more involved than a single `stop_container` call. We must give the underlying application some time to perform graceful shutdown ([example impl](https://medium.com/@dsilverdi/graceful-shutdown-in-go-a-polite-way-to-end-programs-6af16e025549)).

The general idea is that we first make a "polite" call asking the container to be stopped, [giving it a predetermined period of time](https://docs.rs/bollard/latest/bollard/container/struct.StopContainerOptions.html#structfield.t) (e.g. 20 seconds) and then observe the container's status via the `wait` call the other task is performing.

For now, I'd suggest that termination is removed from this PR since it will be a substantial undertaking, which should be separated into its own patch.

Comment (lffg) [reply worker/src/runner/container_rt.rs]: Btw, I think we should use the [`KillContainer` operation](https://docs.docker.com/engine/api/v1.45/#tag/Container/operation/ContainerKill) instead, as it doesn't require us to define a deadline "after which the container is _killed_", unlike `StopContainer`.

Assuming that we use the kill operation instead, we can first make a kill call passing the SIGTERM status and then, after our defined graceful shutdown deadline, make a second kill call, passing SIGKILL. This is advantageous for us since after such a second call we can also emit the _container killed_ `instance::Status` message.

File: worker/src/runner/container_rt.rs (hunk 1 above)
Comment (lemosep) [reply worker/src/runner/container_rt.rs]: `InstanceId` doesn't implement `std::fmt::Display`, may I use `{:#?}`

File: worker/src/sender/mod.rs (hunk 5 above)
Comment (lemosep) [reply worker/src/sender/mod.rs]: I've placed it over `http_path`, any other suggestion?

File: worker/src/runner/container_rt.rs (hunk 1 above)
Comment (lemosep) [reply worker/src/runner/container_rt.rs]: I've decided to implement `fmt::Display` for `InstanceId`, came out better tbh!

File: worker/src/runner/container_rt.rs (hunk 9 above)
Comment (lemosep) [reply worker/src/runner/container_rt.rs]: You're right! the `bollard` crate offers a `option` attribute under `Docker::stop_container`, which contains the Number of seconds to wait before killing the container, allowing us to execute the graceful shutdown. 

With this in mind `kill_container` is already implemented, since no wait time is given to stop execution.

I say we change termination and add `kill_container` as termination is currently being implemented.

File: worker/src/runner/container_rt.rs (hunk 6 above)
Comment (lemosep) [reply worker/src/runner/container_rt.rs]: Yeah, it would be on this `loop` forever, I'll add a validation on `Runner` before even entering the `container_runtime`, to avoid lost tasks.

Also, I am thinking on adding an `attempt_limit` to the loop, maybe 5, wdyt?

File: worker/src/runner/container_rt.rs (hunk 9 above)
Comment (lffg) [reply worker/src/runner/container_rt.rs]: Yep, the problem with `stop_container` is that we don't want Docker to kill the container after the elapsed time. We should kill the container so that we know _we_ killed it and retain control over it.

1. Call `KillContainer` passing SIGTERM.
1. If the container gracefully exited (we would know since we're `wait`ing over that container), we're done; everything ok. Report `instance::Status::Terminated`.
1. Wait for the shutdown deadline.
1. After the deadline elapses, call `KillContainer` again, but passing SIGKILL.
1. When we receive the killed status (also via the `wait` call), we report `instance::Status::Killed`.

But let's leave it to another PR.

File: worker/src/runner/container_rt.rs (hunk 1 above)
Comment (lffg) [reply worker/src/runner/container_rt.rs]: SGTM

File: worker/src/runner/container_rt.rs (hunk 9 above)
Comment (lemosep) [reply worker/src/runner/container_rt.rs]: Wait, so this termination is incomplete then? Should I just remove it from the PR?

Comment (lffg) [reply worker/src/runner/container_rt.rs]: Yes

File: worker/src/runner/container_rt.rs (hunk 6 above)
Comment (lffg) [reply worker/src/runner/container_rt.rs]: We are probably going to need something a little more involved when we implement termination (also, in another PR).

Comment (lffg) [reply worker/src/runner/container_rt.rs]: Also, take a look at exponential backoff retry strategy.

**Review by lffg — APPROVED** [review]
